### DISCUSSION TITLE
Docker를 왜 사용하는가?

### DISCUSSION THREAD

**Discussion by JoisFe** [cs-collections/cs-collections-discussion#26] [h. Docker]
# Docker를 왜 사용하는가?

### 클라우드 환경으로 이주
최근 애플리케이션의 클라우드 환경 이주는 여러 기업, 조직의 매우 큰 관심사이다.
서버, 스토리지, 네트워크, 전원 등 Cloud 서비스 기업에 맡길 수 있기 때문이다.

### 클라우드 환경의 장점
- 무한한 확장성
- 새로운 환경에 애플리케이션 배포의 편리성
- 비용 절감 (실제 사용한 만큼만 부담하면 됨)

### 클라우드 환경 이주의 2가지 방식
1. IaaS(Infrastructure as a Service, 서비스로서의 인프라)
2. PaaS(Platform as a Service, 서비스로서의 플랫폼)

### 두가지 방식의 장점 및 단점
1. IaaS
- 애플리케이션의 각 컴포넌트를 가상 머신에 동작시키게 됨
- 이로 인해 특정 클라우드에 종속되지 않음
- 하지만 운영비가 상승
2. PaaS
- 애플리케이션의 각 컴포넌트를 하나씩 클라우드 매니지드 서비스(managed service)로 옮겨야 하는 까다라운 프로젝트가 필요함
- 애플리케이션이 특정 클라우드에 종속되는 결과를 낳음
- 운영비는 절감할 수 있음

### 도커
- 애플리케이션의 각 컴포넌트를 컨테이너로 이주한 다음 애저 쿠버네티스 서비스 혹은 아마존 일래스틱 컨테이너 서비스 혹은 직접 구축한 도커 클러스터에서 전체 애플리케이션을 실행할 수 있음
- 특정 클라우드에 종속되지 않으므로 원하는 클라우드 서비스나 데이터센터, 심지어 로컬 환경에서도 운영 가능함
- 이로 인해 낮은 운영비, 이식성 확보

### 정리
1. IaaS 모델
- 애플리케이션을 구성하는 각 컴포넌트가 모두 가상머신에서 독립적으로 동작
- 이주 과정은 쉽지만 가상 머신의 성능을 완전히 활용 못함, 운영비 비쌈
2. PaaS 모델
- 애플리케이션을 구성하는 각 컴포넌트가 클라우드 서비스 제공자의 매니지드 서비스에 종속됨
- 운영비가 저렴하고 관리가 쉽지만 이주 과정이 복잡
3. 도커
- 애플리케이션의 각 컴포넌트는 컨테이너에서 동작
- 각 컴포넌트는 가상 머신처럼 독립적이지만 경량이며 PaaS의 매니지드 서비스만큼 효율적
- 도커화된 애플리케이션은 이식성이 뛰어남
- 애저 쿠버네티스 서브시 같은 매니지드 서비스는 물론이고 AWS 혹은 데이터센터로도 별도의 수정없이 이주 가능

## 레거시 애플리케이션 현대화
이전에 말했듯 컨테이너를 활용하면 거의 모든 애플리케이션을 클라우드에서 실행 가능

### 문제점 : 기존의 애플리케이션의 구조를 낡은 모놀리식 설계로 방치한다면...
- 도커 혹은 클라우드 플랫폼의 진가가 발휘되기 어려움
- 모놀리식 설계를 가진 애플리케이션 또한 컨테이너에서 동작하는데는 문제가 되지 않음 그러나 기민성에 제약이 따름

### 레거시 애플리케이션을 도커로 이주하기
애플리케이션을 전면적으로 재구현하지 않고도 새로운 패텅을 도입하는데 매우 간단
1. Dockerfile 스크립트와 도커 컴포즈 문법을 따라 애플리케이션을 단일 컨테이너로 옮김 (이것만으로도 통짜 애플리케이션의 컨테이너 이주로 끝남)
2. 통짜 애플리케이션을 분할해 기능별로 별도의 컨테이너에 배치

컨테이너는 가상 네트워크를 통해 외부에 노출되지 않고도 서로 통신 가능하기 때문에 2.와 같은 일이 가능

### 새롭운 설계로 거듭난 애플리케이션은 MSA의 다양한 장점을 누릴 수 있음
1. 핵심 기능을 작고 독립된 단위로 만들어 따로따로 다루면서 변경 내용을 빠르게 테스트 가능 (애플리케이션의 전체를 변경한 것이 아니라 해당 기능이 포함된 컨테이너만을 변경했기 때문)
2. 해당 기능의 확장성을 조절 가능, 필요에 맞는 적절한 기술 기반 선택 가능

**Comment by ahngilwoong** [comment]
잘 봤습니다. 도커에 대해 무지한 상태인데
MSA로 구축할 경우 한 격리된 컨테이너마다 각각의 서비스(서버)를 만들 수 있다는 점이 맞나요?
또 이 격리된 컨테이너가 하나의 인스턴스(PC?)에 여러개 있으면 이 컨테이너들은 하나의 인스턴스의 자원을 공유하는지와
결국 ip는 같고 port번호가 다른 형식인지가 궁금합니다.
